src/app/qgsmaptoolchamferfillet.cpp

```diff
@@ -0,0 +1,602 @@
+      }
+    }
+    else
+    {
+      value1 = ( value1 + value2 ) / 2.0;
+    }
+
+    applyOperation( value1, value2, e->modifiers() );
+  }
+}
+
+void QgsMapToolChamferFillet::canvasMoveEvent( QgsMapMouseEvent *e )
+{
+  if ( mOriginalGeometry.isNull() || !mRubberBand )
+  {
+    QgsPointLocator::Match match;
+    match = mCanvas->snappingUtils()->snapToCurrentLayer( e->pos(), QgsPointLocator::Types( QgsPointLocator::Vertex ) );
+    mSnapIndicator->setMatch( match );
+    return;
+  }
+
+  // reduce the number of call to max 1 per 100ms
+  long currTime = QDateTime::currentMSecsSinceEpoch();
+  if ( currTime - mLastMouseMove < 100 )
+    return;
+  mLastMouseMove = currTime;
+
+  mGeometryModified = true;
+
+  const QgsPointXY mapPoint = e->mapPoint();
+  mSnapIndicator->setMatch( e->mapPointMatch() );
+
+  double value1, value2;
+  calculateDistances( mapPoint, value1, value2 );
+
+  const QString op = QgsSettingsRegistryCore::settingsDigitizingChamferFilletOperation->value();
+  if ( op == "chamfer" )
+  {
+    if ( e->modifiers() & Qt::ShiftModifier )
+    {
+      value1 = ( value1 + value2 ) / 2.0;
+      value2 = value1;
+    }
+  }
+  else
+  {
+    value1 = ( value1 + value2 ) / 2.0;
+  }
+
+  bool locked = QgsSettingsRegistryCore::settingsDigitizingChamferFilletLock1->value();
+  if ( locked )
+    value1 = QgsSettingsRegistryCore::settingsDigitizingChamferFilletValue1->value();
+
+  locked = QgsSettingsRegistryCore::settingsDigitizingChamferFilletLock2->value();
+  if ( locked )
+    value2 = QgsSettingsRegistryCore::settingsDigitizingChamferFilletValue2->value();
+
+  if ( mUserInputWidget )
+  {
+    mUserInputWidget->blockSignals( true );
+    if ( op == "chamfer" )
+    {
+      mUserInputWidget->setValue1( value1 );
+      mUserInputWidget->setValue2( value2 );
+    }
+    else
+    {
+      mUserInputWidget->setValue1( value1 );
+    }
+    mUserInputWidget->blockSignals( false );
+    mUserInputWidget->setFocus( Qt::TabFocusReason );
+    mUserInputWidget->editor()->selectAll();
+  }
+
+  //create chamfer geometry using geos
+  updateGeometryAndRubberBand( value1, value2 );
+}
+
+void QgsMapToolChamferFillet::prepareGeometry( const QgsPointLocator::Match &match, QgsFeature &snappedFeature )
+{
+  QgsVectorLayer *vl = match.layer();
+  if ( !vl )
+  {
+    return;
+  }
+
+  mOriginalGeometry = QgsGeometry();
+  mManipulatedGeometry = QgsGeometry();
+
+  //assign feature part by vertex number (snap to vertex) or by before vertex number (snap to segment)
+  const QgsGeometry geom = snappedFeature.geometry();
+  if ( geom.isNull() )
+  {
+    return;
+  }
+  mOriginalGeometry = geom;
+
+  const Qgis::WkbType geomType = geom.wkbType();
+  if ( QgsWkbTypes::geometryType( geomType ) != Qgis::GeometryType::Line && QgsWkbTypes::geometryType( geomType ) != Qgis::GeometryType::Polygon )
+    return;
+
+  if ( !match.hasEdge() && !match.hasVertex() )
+    return;
+
+  mManipulatedGeometry = QgsGeometry( geom.constGet()->clone() );
```

Comment:
```suggestion
  mManipulatedGeometry = geom;
```

Reply:
Can you do some more familiarization with the QgsGeometry class please? It's implicitly shared, and already handles detachment for modifications internally.